torch/csrc/distributed/rpc/request_callback_impl.cpp

```diff
@@ -411,6 +413,7 @@ void RequestCallbackImpl::processPythonRRefFetchCall(
       Message m =
           PythonRRefFetchRet(std::move(*result).toIValues()).toMessage();
       m.setId(messageId);
+      rref->waitAllDevices();
```

Comment:
the correct ordering should be:

On processPythonRemoteCall:

1. compute value for the RRef
2. record events and store events in RRef
3. set RRef value, this will unblock 

On processPythonRRefFetchCall:

a. get the RRef value
b. when a is done, we know that 3. is done and hence 2. is done too. So that we can use the cuda events on the RRef.
c. call `ctx-getStream(device)` for each device used by the RRef, and then let RRef's CUDAEvent block the stream returned by `ctx-getStream(device)`. With this, the subsequent TensorPipe pipe->write will use the same streams and hence the ordering will be preservinged, i.e., TensorPipe comm must happen after all pending CUDA ops that contributes to the RRef value.